Port MASTG v1->v2 MASTG-TEST-0002: Testing Local Storage for Input Validation (android) (by @appknox) (#3328)
OWASP/owasp-mstg: tests-beta/android/MASVS-CODE/MASTG-TEST-0288.md

---
title: Use of Local Storage for Input Validation
platform: android

> id: MASTG-TEST-0288

Reviewer: you might need a new id just before merge (test 288 already exists)

type: [static]
weakness: MASWE-0082
profiles: [L1, L2]
---

## Overview

> Data stored in Android's `SharedPreference`s can be tampered with on a rooted device. If an application reads this data without verifying its integrity (e.g., with an HMAC signature), it can lead to security vulnerabilities. This test checks if the application properly validates data read from local storage.

Reviewer: This is already really well-written @jeel38! How about we generalise it a bit and add a note, comment, or intro that this test is not only about SharedPrefs, but also any other local storage, like the v1 test mentions?

Additionally, we should add the requirement for data authenticity (HMAC is about both integrity and authenticity after all):

```suggestion
Data stored in Android's `SharedPreference`s can be tampered with on a rooted device. If an application reads this data without verifying its integrity and authenticity (e.g., with an HMAC signature), it can lead to security vulnerabilities. This test checks if the application properly validates data read from local storage.
```

What we could highlight here to help the reader is that SHA (integrity check without authenticity check) might not be optimal unless the SHA hashes are stored in a backend server. Therefore, by using HMAC, we only need to store the key safely and the data + HMAC hash in the local storage.

## Steps

1. Run a static analysis tool such as @MASTG-TOOL-0110 on the code and look for patterns where data is read from `SharedPreferences` without a corresponding integrity check.

## Observation

The output identifies code where `SharedPreferences` data is loaded without an integrity check.

## Evaluation

> The test fails as the application does not verify the integrity of data loaded from `SharedPreferences`. Without an integrity check like `HMAC`, an attacker with root access can directly edit the SharedPreferences XML file, modifying values to grant themselves higher privileges. The app then reads and acts on this tampered data, leading to a critical security failure, like privilege escalation.

Reviewer: ```suggestion
The test fails if the application does not verify the integrity and authenticity of data loaded from local storage such as `SharedPreferences`.
```

I would consolidate the other sentences as part of the overview to explain the risk etc...

Author: Please Refer to this comment:  https://github.com/OWASP/mastg/pull/3328#discussion_r2326621776

Reviewer: To my understanding, @cpholguera pointed out that suggestion for the DEMO, this is the TEST file.
Reading this TEST's Overview and Evaluation sections seems to be some repetition.
